### DISCUSSION TITLE
JanetRS - Janet high-level bindings for Rust: version 0.7.0

### DISCUSSION THREAD

**Discussion by GrayJack** [janet-lang/janet#1426] [Show and tell]
I just released a new version of JanetRS. In this one, the changelog is more on the small side, so here it is

## Changes

- **BREAKING:** Refactor: Rename `JanetArgs::get_unwraped` to `JanetArgs::get_unwrapped`
- **BREAKING:** Refactor: Rename `FileFlags::is_serializeble` to `FileFlags::is_serializable`
- Feat: Add `JanetAbstract::take` method
- Feat: Add `JanetAbstract::into_inner` method
- Feat: Add `JanetTuple::{sourcemap, set_sourcemap}` methods
- Refactor: Simplify `JanetAbstract::new`
- Fix: Fix undefined behavior in `JanetArray::as_{ref,mut}` methods
- Fix: Fix clippy lints
- docs: Update a few documentation comments

I have been off for a couple of months now, so I intend to see what changed in the C public API, add the methods, and make improvements to the library, but I wanted to release the UB fix and the breaking changes soon.

**Comment by ghost** [comment]
How is integrating with a GC from Rust? I've heard it's a real headache unless everything is `unsafe`. I happen to know there are no fewer than three Lua binding crates due to philosophical differences on the matter.

**Reply by GrayJack** [reply]
What exactly do you mean by integrating with a GC from Rust? Do you mean making an API to talk directly with the GC or do you mean about the GC-handled objects?

 For both of them, the Janet public C API is very nice, which means it is easier to use correctly and that helps a lot.

For the first, I only exposed the APIs for forcing collecting which is unsafe. For rooting a value or locking the GC, I used the guard pattern to undo at the end of the scope, which is what the user wants most of the time. You can always wrap the guard in `ManuallyDrop` if you want to extend the scope.

For the second, as far as I'm aware, as long the object is reachable, it will be alive, so unless you modify the internal `GcObject`, get the object directly from the memory address somehow without the GC knowing you got the object or something else that would already be unsafe to do in Rust, the object will always be alive for the scope that you created the instance. So for library creation (the thing I used and stressed the library the most), if your function receives a Janet type and the Janet types you create inside it will always be alive for the function scope.

That doesn't mean that something wrong with my knowledge of the Janet GC can't happen (I'm not even close to being an expert in that field).
That doesn't also mean that there may be other ways to deal with the GCed objects as well.

I hope that answer your question
